### PR TITLE
Only inspect user-written predicates for privacy concerns

### LINKED ISSUE #123288: regression: type is now private?
```
[INFO] [stdout] error: type `bright_lang_lexer::SingleLineCommentToken` is private
[INFO] [stdout]   --> /opt/rustwide/cargo-home/registry/src/index.crates.io-6f17d22bba15001f/logos-0.11.4/src/lexer.rs:13:5
[INFO] [stdout]    |
[INFO] [stdout] 13 |     source: &'source Token::Source,
[INFO] [stdout]    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ private type
```

* https://crater-reports.s3.amazonaws.com/beta-1.78-3/beta-2024-03-30/gh/theidiotmachine.bright-lang/log.txt

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #123288

Previously we looked at the elaborated predicates, which, due to adding various bounds on fields, end up requiring trivially true bounds. But these bounds can contain private types, which the privacy visitor then found and errored about.

### CODE REVIEW

File: tests/ui/privacy/generic_struct_field_projection.rs [hunk 1]
```diff
@@ -0,0 +1,27 @@
+//@ check-pass
```
Comment (compiler-errors) [inline tests/ui/privacy/generic_struct_field_projection.rs:1]: This test needs to actually explain what it's testing.

**Review by compiler-errors — CHANGES_REQUESTED** [review]
What predicate ends up mentioning a private type? Specifically, what does `Foo` have to do with `T::Assoc`?

**Comment by compiler-errors** [comment]
> due to adding various bounds on fields

I think it would be helpful if you explained what bounds are being added specifically, and why that's problematic.

**Comment by oli-obk** [comment]
Explained the test and its problematic bounds.

**Comment by wesleywiser** [comment]
@bors p=1

Fixes P-high issue

**Comment by apiraino** [comment]
Beta backport accepted as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-04-11/near/432715133). A backport PR will be authored by the release team at the end of the current development cycle.

@rustbot label +beta-accepted

**Pull request merged**